### ISSUE TITLE
Feature request- Refactoring of Module Variable navigation

### ISSUE THREAD

**Issue by kitos9112** [aztfmod/terraform-azurerm-caf#1696]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Hi!

I began taking a look at this module quite recently and the main issue I'm facing is just understanding the variable type correlation between the root module and its children. 

Below there is a mental process of creating new AAD objects, especially a simple AAD group. 

Firstly, you'd start having to populate this variable with some sort of map data after taking a look at some examples.

```hcl
## Azure AD
variable "azuread" {
  description = "Configuration object - Azure Active Directory resources"
  default     = {}
}
```

Then, you will potentially jump into the `local` equivalent (`azuread`) and find out about other child types.

```hcl
  azuread = {
    azuread_administrative_unit_members = try(var.azuread.azuread_administrative_unit_members, {})
    azuread_administrative_units        = try(var.azuread.azuread_administrative_units, {})
    azuread_api_permissions             = try(var.azuread.azuread_api_permissions, {})
    azuread_applications                = try(var.azuread.azuread_applications, {})
    azuread_apps                        = try(var.azuread.azuread_apps, {})
    azuread_credential_policies         = try(var.azuread.azuread_credential_policies, {})
    azuread_credentials                 = try(var.azuread.azuread_credentials, {})
    azuread_groups                      = try(var.azuread.azuread_groups, {})
    azuread_groups_membership           = try(var.azuread.azuread_groups_membership, {})
    azuread_roles                       = try(var.azuread.azuread_roles, {})
    azuread_service_principal_passwords = try(var.azuread.azuread_service_principal_passwords, {})
    azuread_service_principals          = try(var.azuread.azuread_service_principals, {})
    azuread_users                       = try(var.azuread.azuread_users, {})
  }
```

Secondly, you'd stare at `azuread_groups.tf` because there are other possible local values to get your head around.

```hcl
module "azuread_groups" {
  source   = "./modules/azuread/groups"
  for_each = local.azuread.azuread_groups

  global_settings = local.global_settings
  azuread_groups  = each.value
  tenant_id       = local.client_config.tenant_id
  client_config   = local.client_config
  remote_objects = {
    azuread_administrative_units = local.combined_objects_azuread_administrative_units
  }
}
``` 

If you made it till here, you'd still need to keep reading more internal abstractions (locals, variables, etc...) to understand how to populate this variable present in the child module (`azuread/groups`)

```hcl
# variables.tf
variable "azuread_groups" {
  description = "Set of groups to be created."
}

 # main.tf
resource "azuread_group" "group" {

  administrative_unit_ids = can(var.azuread_groups.administrative_unit_ids) || can(var.azuread_groups.administrative_units) == false ? try(var.azuread_groups.administrative_unit_ids, null) : local.administrative_unit_ids
  assignable_to_role      = try(var.azuread_groups.assignable_to_role, null)
  display_name            = var.global_settings.passthrough || try(var.azuread_groups.global_settings.passthrough, false) == true ? format("%s", local.display_name) : format("%s%s", try(format("%s-", var.global_settings.prefixes.0), ""), local.display_name)
  description             = lookup(var.azuread_groups, "description", null)
  prevent_duplicate_names = lookup(var.azuread_groups, "prevent_duplicate_names", null)
  owners                  = length(local.owners) > 0 ? local.owners : null
  // Note: This module is effected by these bugs:
  // https://github.com/hashicorp/terraform-provider-azuread/issues/464
  // https://github.com/microsoftgraph/msgraph-metadata/issues/92
  // tldr: If your group is initially owned by a service principal and you add a user to the owners, you are not able to remove the user from the owners again. At least one user has to stay owner.
  security_enabled  = try(var.azuread_groups.security_enabled, true)
  visibility        = try(var.azuread_groups.visibility, null)
  mail_enabled      = try(var.azuread_groups.mail_enabled, null)
  writeback_enabled = try(var.azuread_groups.writeback_enabled, null)

}
```

Finally, you could interpret how to populate the initial variable (`var.azuread.azuread_groups`) - Although you're losing some of the static type validations Terraform does by default at initialisation.

In conclusion, it is quite difficult to get your head around all possible variable/local/module combinations. I really like your approach to abstract Terraform's internal complexities, especially when it comes to combining all three main TF providers (azureAD, azureRM, and AzureCAF) 

A couple of questions?

 - Are there any documented logic you guys follow when using the module for non-Rover users?
 - Is there a plan to improve this using `optional` attributes? 

I'm aware of some limitations of this solution (https://github.com/hashicorp/terraform/issues/33450)

### New or Affected Resource(s

azuread_group

### References

https://github.com/aztfmod/terraform-azurerm-caf/issues/461
